File: CSProj/CS650/Lab 3/GraphCut/kmean.cpp
#include "kmean.h"
#include <iostream>


KMean::KMean(int data_dim, int cluster_num)
{
    mDataDim = data_dim;
    mClusterNum = cluster_num;

    mMeans = new double*[mClusterNum];
    for (int i = 0; i < mClusterNum; i++)
    {
        mMeans[i] = new double[mDataDim];
    }

    mMaxIterationCnt = 100;
    mStopError = 0.001;
}

KMean::~KMean()
{
    for (int i = 0; i < mClusterNum; i++)
    {
        delete[] mMeans[i];
    }
    delete[] mMeans;
}

int * KMean::cluster(double * data, int data_size)
{
    int * labels = new int[data_size];

    double * x = new double[mDataDim];
    int label = -1;
    int iterationCnt = 0;
    double lastCost = 0;
    double currentCost = 0;
    int unchanged = 0;
    bool loop = true;
    int* counts = new int[mClusterNum];
    double** next_means = new double*[mClusterNum];	// New model for re-estimation
    for (int i = 0; i < mClusterNum; i++)
    {
        next_means[i] = new double[mDataDim];
    }

    while (loop)
    {
        memset(counts, 0, sizeof(int) * mClusterNum);
        for (int i = 0; i < mClusterNum; i++)
        {
            memset(next_means[i], 0, sizeof(double) * mDataDim);
        }

        lastCost = currentCost;
        currentCost = 0;

        // Classification
        for (int i = 0; i < data_size; i++)
        {
            for(int j = 0; j < mDataDim; j++)
                x[j] = data[i*mDataDim+j];

            currentCost += getLabel(x, &label);

            counts[label]++;
            for (int d = 0; d < mDataDim; d++)
            {
                next_means[label][d] += x[d];
            }
        }
        currentCost /= data_size;

        // Re-estimation
        for (int i = 0; i < mClusterNum; i++)
        {
            if (counts[i] > 0)
            {
                for (int d = 0; d < mDataDim; d++)
                {
                    next_means[i][d] /= counts[i];
                }
                memcpy(mMeans[i], next_means[i], sizeof(double) * mDataDim);
            }
        }

        // Terminal conditions
        iterationCnt++;
        if (fabs(lastCost - currentCost) < mStopError * lastCost)
        {
            unchanged++;
        }
        if (iterationCnt >= mMaxIterationCnt || unchanged >= 3)
        {
            loop = false;
        }
    }

    // Output the label file
    for (int i = 0; i < data_size; i++)
    {
        for(int j = 0; j < mDataDim; j++)
        {
            x[j] = data[i*mDataDim+j];
        }
        getLabel(x, &label);
        labels[i] = label;
    }
    delete[] counts;
    delete[] x;
    for (int i = 0; i < mClusterNum; i++)
    {
        delete[] next_means[i];
    }
    delete[] next_means;


    return labels;
}

void KMean::init(double *data, int data_size)
{
    double* sample = new double[mDataDim];

    for (int i = 0; i < mClusterNum; i++)
    {
        int select = i * data_size / mClusterNum;
        for(int j = 0; j < mDataDim; j++)
            sample[j] = data[select*mDataDim+j];
        memcpy(mMeans[i], sample, sizeof(double) * mDataDim);
    }

    delete[] sample;
}

double KMean::getLabel(const double* sample, int* label)
{
    double dist = -1;
    for (int i = 0; i < mClusterNum; i++)
    {
        double temp = getDistance(sample, mMeans[i]);
        if (temp < dist || dist == -1)
        {
            dist = temp;
            *label = i;
        }
    }
    return dist;
}

double KMean::getDistance(const double* p, const double* q)
{
    double temp = 0;
    for (int d = 0; d < mDataDim; d++)
    {
        temp += (p[d] - q[d]) * (p[d] - q[d]);
    }
    return sqrt(temp);
}

File: CSProj/CS650/Lab 4/kmean.py
'''
Created on Oct 30, 2014

@author: daqing_yi
'''

import numpy as np

class KMeanCluster(object):

    def __init__(self, data_dim, cluster_num, max_iter=1000, stop_err=0.00001):
        self.data_dim = data_dim
        self.cluster_num = cluster_num
        self.max_iter = max_iter
        self.stop_err = stop_err
        
    
    def cluster(self, data):
        self.data = data
        self.data_size = len(data)
        self.labels = np.zeros(self.data_size, np.int)
        
        self.means = []
        for i in range(self.cluster_num):
            self.means.append(np.zeros(self.data_dim))
        for i in range(self.cluster_num):
            select = int(i * self.data_size / self.cluster_num)
            for j in range(self.data_dim):
                self.means[i][j] = self.data[select][j]
                
        next_sums = []
        for j in range(self.cluster_num):
            next_sums.append(np.zeros(self.data_dim, np.double))
        counts = np.zeros(self.cluster_num, np.int)
        iter_cnt = 0
        current_cost = 0.0
        unchanged = 0
        
        loop = True
        while loop==True:
            
            last_cost = current_cost
            current_cost = 0.0
            # Classification
            for i in range(self.data_size):
                cost, l = self.getLabel(self.data[i])
                current_cost += cost
                counts[l] += 1

                for d in range(self.data_dim):
                    next_sums[l][d] += self.data[i][d]
                    
            current_cost = current_cost / self.data_size
            
            # Re-estimation
            for j in range(self.cluster_num):
                if counts[j] > 0:
                    for d in range(self.data_dim):
                        self.means[j][d] = next_sums[j][d] / counts[j]
            
            # Terminal condition
            iter_cnt += 1
            if np.abs(last_cost - current_cost) < self.stop_err * last_cost:
                unchanged += 1
            if iter_cnt > self.max_iter or unchanged >= 3:
                loop = False
        
    
    def getMean(self, idx):
        return self.means[idx]
    
    def getLabel(self, p):
        dist = -1
        label = None
        for i in range(self.cluster_num):
            temp = self.getDistance(p, self.means[i])
            if temp < dist or dist == -1:
                dist = temp
                label = i
        return dist, label

    
    def getDistance(self, p, q):
        
        dist = 0.0
        for j in range(self.data_dim):
            dist += (p[j] - q[j])**2
        return np.sqrt(dist)
    

        